### DISCUSSION TITLE
Proper way of handling per user access to certain records

### DISCUSSION THREAD

**Discussion by Y4shin** [pocketbase/pocketbase#688] [Q&A] [answered]
Hey all,
I am rather new to pocketbase and I want to create a database schema in pocketbase that allows me to control access to a collection by giving individual users rights to access individual records. Here an example:

I have a table `campaign` and a table `post`. Each Post belongs to a certain campaign and a campaign belongs to a certain user:

```
+----------------+
|    campaign    |
+-------+--------+
| title | string |
| owner | User   |
+-------+--------+
```
```
+-------------------------------+
|              post             |
+----------+--------------------+
| title    | string             |
| details  | string             |
| campaign | Relation(campaign) |
+----------+--------------------+
```

So my thought was to create a junction table between `user`and `post` to be able to give individual access to a certain post. Here is the junction table:
```
+--------------------------------+
|           post_access          |
+------+-------------------------+
| user | User                    |
| post | Relation(post)          |
| type | MultipleChoice('read',  |
|      |   'write', 'readwrite') |
+------+-------------------------+
```

I then used an access rule like this to limit access (this is for the view permission):
```sql
(@request.user.id = campaign.owner) ||
(@collection.post_access.post.id = id &&
 @collection.post_access.user = @request.user.id &&
 (@collection.post_access.access_type = "read" ||
  @collection.post_access.access_type = "readwrite"))
```

This seems to work from my testing, had to do some figuring out because I did not really understand the `@collection` directive in these api rules. The way I understand it now is that if I have multiple references to one certain collection it looks for the existence of an record that matches all the statements that I made regarding that collection. So in this example for that statement above to be true either the user needs to be the owner of that campaign or there needs to be a record in the `post_access`collection that connects the user with this post and grants him either `readwrite` or `read` access.

I think that this is a suitable way of handling this kind of access control, however I am not sure of the performance impact of more complex api rules. Therefore my question would be whether there is some other way of doing this, that may be more simple. I just kind of have this feeling that I may have overlooked something while coming up with this mechanism.

**Comment by ganigeorgiev** [comment, marked as answer]
I think that your usage and understanding of the `@collection.*` rule is correct.

Internally we are making a left join for each new `@collection.someCollection` statement. So in other words, the above filter would be roughly translated to the following SQL statement:

```sql
SELECT DISTINCT post.*
FROM post
LEFT JOIN campaign on campaign.id = post.campaign
LEFT JOIN post_access
WHERE post.id = [POST_ID] AND (
  [REQUEST_USER_ID] = campaign.owner OR (
    post_access.post = post.id AND
    post_access.user = [REQUEST_USER_ID] AND
    (post_access.access_type = "read" OR post_access.access_type = "readwrite")
  )
)
LIMIT 1
```

_Actually, you'll have one extra join because of the nested `@collection.post_access.post.id`, but since in your case `post_access.post` is a single relation field, you can optimize a little your rule and directly compare the field to the `id` (I may consider automating this), eg:_
```sql
(@request.user.id = campaign.owner) ||
(@collection.post_access.post = id &&
 @collection.post_access.user = @request.user.id &&
 (@collection.post_access.access_type = "read" ||
  @collection.post_access.access_type = "readwrite"))
```

About the performance - it depends how big is your data set. PocketBase 0.7.4+ comes with a lot of improvements to the relation query look ups - from ~15s to ~0.02s (tested with 2 collections with 10k records each, for both single and multiple relation fields). And this is without cache or additional indexes. In the near future we will have support for defining indexes from the Admin UI, so you'll be able to tune further your query when needed.

Based on your data structure and requirements, at least at the moment I can't think of other more simpler way to handle it.

**Reply by Y4shin** [reply]
Thank you for the speedy reply. The explanation made the `@collection` directive much clearer and I can more confidently use the `@collection` directive in my api rules. Also the tip with the extra join due to the `post.id` is very helpful and clear once one understands the mechanics of `@collection`.

I think that this could be referenced somewhat in the Documentation on API Rules. Maybe something like this:

* API Rules: "Any given rule restricts the access to the corresponding action on any given document(s) to the documents that satisfy the rule under the specific request context (i.e. logged in user) of the API call." (I would add this to the API Rules segment)
* API Rule Syntax: "`@collection.*`filter fields - you can use them to target other collections that the current collection has no direct access to, i.e. that the current collection has no relation field pointing to it. When using `@collection.` fields you can think of it as such: for for any given request context to be given access to a record on the current collection there needs to be a record on the collection referenced by `@collection.*` that satisfies the logic laid out by the rule. When referencing the same collection with `@collection.*` multiple times, like in the example below, both `@collection.*` instances refer to the same record on that given collection. For those of you familiar with SQL, `@collection.*` is internally implemented with a left join." (this would replace the explanation of `@collection` in the current documentation)

Just to be clear, this is just something that I thought of of the top of my head to more clearly communicate the nature of the `@collection.*` as I kind of figured that it would be something like this when I looked at the docs, but it was not explicitly stated (or I did not understand it properly).
